### PR TITLE
Add batch subgroup verification, batched bucket addition, parallelisation

### PR DESCRIPTION
**Pull request by jon-chuang**
Verification for 2 million BW6 points: 6-7s on 8 cores.

### CODE REVIEW

File: algebra-core/src/curves/batch_verify.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+// We get the greatest power of 2 number of buckets
+// such that we minimise the number of rounds
+// while satisfying the constraint that number of rounds * buckets * 2 < n
+
+// Number of buckets is always greater than new security param.
+// So only need 1 round subsequently
+fn get_max_bucket(security_param: usize, n_elems: usize) -> (usize, usize) {
+    let mut log2_num_buckets = 1;
+    let num_rounds =
+        |log2_num_buckets: usize| -> usize { (security_param - 1) / log2_num_buckets + 1 };
```
Comment (psivesely) [inline algebra-core/src/curves/batch_verify.rs:133]: This should be `ceil(security_param / (log2_num_buckets - 1))`, which is not the same as `security_param / (log2_num_buckets - 1) + 1` when `log2_num_buckets - 1` divides `security_param`.

Comment (psivesely) [reply algebra-core/src/curves/batch_verify.rs:133]: You'll need to initialize `log2_num_buckets = 2`.

**Comment by psivesely** [comment]
As we discussed on Slack, let’s switch to a non-recursive bucket test, at least as a first step. I don’t think your `get_num_buckets` is optimizing m. Let n be the number of group elements we wish to check subgroup membership for, s be the security parameter, and r be the ratio of the cost of a subgroup membership test to a group addition. The formula I derived for finding m=log<sub>2</sub>(M) is to find the value 1 ≤ m ≤ floor(log<sub>2</sub>(n)) that minimizes ceil(s/m) * (n + Mr).

What's the smallest cofactor for BW6? Just trying to figure out if it's worth it to use the small exponent test.

**Comment by psivesely** [comment]
Here's the equation I derived for optimizing the recursive bucket test, where d≥2 is the recursion depth and 1 ≤ m<sub>d</sub> ≤ ... ≤ m<sub>1</sub> ≤ floor(log2(n)). Just in case we want to go this direction later.

Edit: Updated equation. Jon and I have a still have a slight disagreement on the number of rounds necessary to get s bits security, but are going to work it out over a call.

**Comment by jon-chuang** [comment]
TODO: pass an `Rng` parameter for batch verify bucket assignment randomness, and use fiat-shamir for the seed for `no_std` only as a compromise, retaining `thread_rng` for normal use. Need to split seed into multiple new seeds to use on multiple threads, and need to make sure this methodology is correct. 

References:

**Comment by jon-chuang** [comment]
Reference doc of current progress courtesy of Psi
Screenshot from 2020-08-25 10-02-22

**Comment by psivesely** [comment]
Deleted the reference doc @jon-chuang because it had errors as I pointed out in Slack. I still believe the cost equation above is guaranteed to give us more than lambda bits security. The (1+r) factor should be actually be replaced with just r, but r is like 280 for us, so this won't make much of a difference. Happy to copy my thoughts over from Slack justifying the cost equation above.

Anyway, determining the cost function for lambda bits security is for future reference only, as we decided to go with the non-recursive version for this initial PR and circle back on the recursive version if we have time before the trusted setup ceremony begins.

**Comment by kobigurk** [comment]
Can this be closed?

**Pull request closed without merging**